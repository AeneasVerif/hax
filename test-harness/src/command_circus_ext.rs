use lazy_static::{__Deref, lazy_static};
use std::{ffi::OsStr, process::Command};

pub trait CommandCircusExt {
    fn circus<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Self;
}

impl CommandCircusExt for Command {
    fn circus<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Command {
        use assert_cmd::cargo::cargo_bin;
        use std::path::PathBuf;
        struct Paths {
            driver: PathBuf,
            engine: PathBuf,
            cargo_circus: PathBuf,
        }
        const CARGO_CIRCUS: &str = "cargo-circus";
        lazy_static! {
            static ref PATHS: Option<Paths> = {
                if let "yes" | "y" | "true" | "1" = std::env::var("CARGO_TESTS_ASSUME_BUILT").unwrap_or("".into()).to_lowercase().as_str() {
                    return None;
                }
                let root = std::env::current_dir().unwrap();
                let root = root.parent().unwrap();
                let engine_dir = root.join("engine");
                // Make sure binaries are built
                assert!(Command::new("cargo")
                        .args(&["build", "--workspace", "--bins"])
                        .status()
                        .unwrap()
                        .success());
                assert!(Command::new("dune")
                        .args(&["build"])
                        .env("CIRCUS_JSON_SCHEMA_EXPORTER_BINARY", cargo_bin("circus-export-json-schemas"))
                        .current_dir(engine_dir.clone())
                        .status()
                        .unwrap()
                        .success());
                Some(Paths {
                    driver: cargo_bin("driver-circus-frontend-exporter"),
                    cargo_circus: cargo_bin(CARGO_CIRCUS),
                    engine: engine_dir.join("_build/install/default/bin/circus-engine"),
                })
            };
        }
        let mut cmd = match PATHS.deref() {
            Some(paths) => {
                let mut cmd = Command::new(paths.cargo_circus.clone());
                cmd.env("CIRCUS_RUSTC_DRIVER_BINARY", paths.driver.clone());
                cmd.env("CIRCUS_ENGINE_BINARY", paths.engine.clone());
                cmd
            }
            None => Command::new(CARGO_CIRCUS),
        };
        cmd.args(args);
        // As documented in
        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths,
        // [cargo run] (and thus also [cargo test]) sets dynamic
        // library paths, which causes some issues with dependencies
        // when compiling without rustup
        for env in ["DYLD_FALLBACK_LIBRARY_PATH", "LD_LIBRARY_PATH"] {
            cmd.env_remove(env);
        }
        cmd
    }
}
