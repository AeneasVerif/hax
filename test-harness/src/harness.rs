mod command_circus_ext;
use command_circus_ext::*;
use serde_json::Value;
use std::process::{Command, Stdio};

#[derive(Clone, Debug)]
pub enum TestKind {
    Translate { backend: String },
    Lint { linter: String },
}

impl TestKind {
    fn as_subcommands(&self) -> [String; 2] {
        match self {
            TestKind::Lint { linter } => ["lint".to_string(), linter.clone()],
            TestKind::Translate { backend } => ["into".to_string(), backend.clone()],
        }
    }
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSpec {
    /// a slow test can be made optional
    pub optional: bool,
    /// a broken test a test that should succeed (or fail) but does
    /// not dues to a bug to be fixed (see field [issue_id] below)
    pub broken: bool,
    /// Github issue ID
    pub issue_id: Option<u64>,
    /// Is that a positive or a negative test?
    pub positive: bool,
}
impl From<Value> for TestSpec {
    /// Parse a JSON value into a TestSpec
    fn from(o: Value) -> Self {
        let as_bool = |key: &str, default: bool| {
            let b = &o[key];
            if b.is_null() {
                default
            } else {
                b.as_bool().unwrap()
            }
        };
        Self {
            optional: as_bool("optional", false),
            positive: as_bool("positive", true),
            broken: as_bool("broken", false),
            issue_id: o["issue_id"].as_u64(),
        }
    }
}

/// The information for a test is given by `cargo metadata`
#[derive(Clone, Debug)]
pub struct TestInfo {
    pub name: String,
    pub manifest: std::path::PathBuf,
    pub description: Option<String>,
}

#[derive(Clone, Debug)]
pub struct Test {
    pub kind: TestKind,
    pub info: TestInfo,
    pub spec: TestSpec,
}

impl std::fmt::Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} - {:?}", self.info.name, self.kind)?;
        if let Some(issue_id) = self.spec.issue_id {
            write!(f, " #{}", issue_id)?;
        };
        Ok(())
    }
}

impl Test {
    fn into_runner(self, workspace: String) -> Result<(), libtest_mimic::Failed> {
        // 1. cook a command
        let mut cmd = Command::circus(&["-C"]);
        cmd.arg("--manifest-path").arg(self.info.manifest.clone());
        cmd.arg(";");
        cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
        cmd.args(self.kind.as_subcommands());

        // 2. execute it (twice, idea of @franziskuskiefer, so that
        // the messages related to building dependencies are not
        // included in the second one)
        let _ = cmd.output().unwrap();
        let out = cmd.output().unwrap();

        let command_successful = out.status.success();
        let cleanup = |s: String| {
            s.replace(r"\", "/")
                .replace(&workspace, "WORKSPACE_ROOT")
                .replace("Blocking waiting for file lock on build directory", "")
                .trim()
                .to_string()
        };
        let serr = cleanup(String::from_utf8_lossy(&out.stderr).to_string());
        let sout = cleanup(String::from_utf8_lossy(&out.stdout).to_string());

        // 3. make sure the test is successful
        let err = |s: &str| {
            Err(format!(
                "Command {s}.\nThe command was: {:?}{}",
                cmd,
                if command_successful {
                    "".to_string()
                } else {
                    format!("\nSTDOUT:\n{}\nSTDERR:\n{}", sout, serr)
                }
            ))
        };
        match (command_successful, (self.spec.positive, self.spec.broken)) {
            (false, (false, false) | (true, true)) => {
                #[derive(serde::Serialize)]
                struct Output {
                    serr: String,
                    sout: String,
                    exit: i32,
                }
                let exit = out.status.code().unwrap_or(std::i32::MAX);
                let out = &Output { serr, sout, exit };
                let name = format!(
                    "{}+{}",
                    self.info.name,
                    self.kind.as_subcommands().join("-")
                );
                insta::with_settings!({
                    info => &self.spec,
                }, { Ok(insta::assert_toml_snapshot!(name, out)) })
            }
            (false, (false, true)) => err("failed, but this is a negative test marked broken")?,
            (false, (true, false)) => err("failed")?,
            (true, (true, false) | (false, true)) => Ok(()),
            (true, (true, true)) => err("succeeded, but this is a positive test marked broken")?,
            (true, (false, false)) => err("succeeded, but this is a negative test")?,
        }
    }

    fn into_trial(&self, workspace: &String) -> libtest_mimic::Trial {
        libtest_mimic::Trial::test(format!("{}", &self), {
            let test = self.clone();
            let workspace = workspace.clone();
            move || test.clone().into_runner(workspace)
        })
        .with_kind(if self.spec.positive {
            "positive"
        } else {
            "negative"
        })
        .with_ignored_flag(self.spec.optional)
    }
}

/// Given [metadata] the table declared in a test's [Cargo.toml]
/// [workspace.circus-tests], this function returns a list of tests
fn parse_circus_tests_metadata(info: TestInfo, metadata: &Value) -> Vec<Test> {
    metadata
        .as_object()
        .expect("Expected value at key [circus-tests] to be a dictionary")
        .into_iter()
        .flat_map(|(a, o)| {
            o.as_object()
                .expect(format!("Expected value at key [{}] be a dictionary", a).as_str())
                .into_iter()
                .flat_map(|(key, o)| key.split("+").map(|k| (k.trim().to_string(), o.clone())))
                .map(|(b, o)| (a.clone(), b, o))
        })
        .map(|(a, b, o)| Test {
            spec: o.into(),
            info: info.clone(),
            kind: match a.as_str() {
                "into" => TestKind::Translate { backend: b },
                "lint" => TestKind::Lint { linter: b },
                _ => panic!("unexpected metadata [circus-tests.{}.{}]", a, b),
            },
        })
        .collect()
}

fn main() {
    let metadata = cargo_metadata::MetadataCommand::new()
        .manifest_path("../tests/Cargo.toml")
        .exec()
        .unwrap();
    let workspace_root: String = metadata.workspace_root.into();

    libtest_mimic::run(
        &libtest_mimic::Arguments::from_args(),
        metadata
            .packages
            .into_iter()
            .flat_map(|o| {
                parse_circus_tests_metadata(
                    TestInfo {
                        name: o.name,
                        description: o.description,
                        manifest: o.manifest_path.into(),
                    },
                    &o.metadata["circus-tests"],
                )
            })
            .map(|test| test.into_trial(&workspace_root))
            .collect(),
    )
    .exit();
}
