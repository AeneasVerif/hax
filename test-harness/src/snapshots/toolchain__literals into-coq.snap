---
source: test-harness/src/harness.rs
expression: snapshot
info:
  kind:
    Translate:
      backend: coq
  info:
    name: literals
    manifest: literals/Cargo.toml
    description: ~
  spec:
    optional: false
    broken: false
    issue_id: ~
    positive: true
    snapshot:
      stderr: true
      stdout: true
    include_flag: ~
    backend_options: ~
---
exit = 1
stderr = """
Finished `dev` profile [unoptimized + debuginfo] target(s) in XXs
\u001B[1m\u001B[91merror\u001B[0m: \u001B[1m[HAX0001] (AST import) something is not implemented yet.This is discussed in issue https://github.com/hacspec/hax/issues/156.
Please upvote or comment this issue if you see this error message.
Sorry, Hax does not support declare-first let bindings (see https://doc.rust-lang.org/rust-by-example/variable_bindings/declare.html) for now.\u001B[0m
  \u001B[1m\u001B[94m-->\u001B[0m literals/src/lib.rs:43:21
\u001B[1m\u001B[94m   |\u001B[0m
\u001B[1m\u001B[94m43 |\u001B[0m #[derive(PartialEq, Eq)]
\u001B[1m\u001B[94m   |\u001B[0m\u001B[1m\u001B[91m                     ^^\u001B[0m
\u001B[1m\u001B[94m   |\u001B[0m"""
[[stdout.diagnostics]]
message = '''
(AST import) something is not implemented yet.This is discussed in issue https://github.com/hacspec/hax/issues/156.
Please upvote or comment this issue if you see this error message.
Sorry, Hax does not support declare-first let bindings (see https://doc.rust-lang.org/rust-by-example/variable_bindings/declare.html) for now.'''
spans = ['Span { lo: Loc { line: 43, col: 20 }, hi: Loc { line: 43, col: 22 }, filename: Real(LocalPath("literals/src/lib.rs")), rust_span_data: None }']

[stdout.files]
"Literals.v" = '''
(* File automatically generated by Hacspec *)
From Coq Require Import ZArith.
Require Import List.
Import List.ListNotations.
Open Scope Z_scope.
Open Scope bool_scope.
Require Import Ascii.
Require Import String.
Require Import Coq.Floats.Floats.
From RecordUpdate Require Import RecordSet.
Import RecordSetNotations.
From Core Require Import Core.



(* NotImplementedYet *)



Definition math_integers (x : t_Int) `{andb (f_gt (x) (impl_Int__e_unsafe_from_str (("0"%string : string)))) (f_lt (x) (impl_Int__e_unsafe_from_str (("16"%string : string)))) = true} : t_u8 :=
  let _ : t_Int := f_lift ((3 : t_usize)) in
  let e_neg_dec := impl_Int__e_unsafe_from_str (("-340282366920938463463374607431768211455000"%string : string)) in
  let e_pos_dec := impl_Int__e_unsafe_from_str (("340282366920938463463374607431768211455000"%string : string)) in
  let e_neg_hex := impl_Int__e_unsafe_from_str (("-340282366920938463463374607431768211455000"%string : string)) in
  let e_pos_hex := impl_Int__e_unsafe_from_str (("340282366920938463463374607431768211455000"%string : string)) in
  let e_neg_octal := impl_Int__e_unsafe_from_str (("-340282366920938463463374607431768211455000"%string : string)) in
  let e_pos_octal := impl_Int__e_unsafe_from_str (("340282366920938463463374607431768211455000"%string : string)) in
  let e_neg_bin := impl_Int__e_unsafe_from_str (("-340282366920938463463374607431768211455000"%string : string)) in
  let e_pos_bin := impl_Int__e_unsafe_from_str (("340282366920938463463374607431768211455000"%string : string)) in
  let _ := f_gt (impl_Int__e_unsafe_from_str (("-340282366920938463463374607431768211455000"%string : string))) (impl_Int__e_unsafe_from_str (("340282366920938463463374607431768211455000"%string : string))) in
  let _ := f_lt (x) (x) in
  let _ := f_ge (x) (x) in
  let _ := f_le (x) (x) in
  let _ := f_ne (x) (x) in
  let _ := f_eq (x) (x) in
  let _ := f_add (x) (x) in
  let _ := f_sub (x) (x) in
  let _ := f_mul (x) (x) in
  let _ := f_div (x) (x) in
  let _ : t_i16 := impl_Int__to_i16 (x) in
  let _ : t_i32 := impl_Int__to_i32 (x) in
  let _ : t_i64 := impl_Int__to_i64 (x) in
  let _ : t_i128 := impl_Int__to_i128 (x) in
  let _ : t_isize := impl_Int__to_isize (x) in
  let _ : t_u16 := impl_Int__to_u16 (x) in
  let _ : t_u32 := impl_Int__to_u32 (x) in
  let _ : t_u64 := impl_Int__to_u64 (x) in
  let _ : t_u128 := impl_Int__to_u128 (x) in
  let _ : t_usize := impl_Int__to_usize (x) in
  impl_Int__to_u8 (f_add (x) (f_mul (x) (x))).

Definition panic_with_msg '(_ : unit) : unit :=
  never_to_any (panic_fmt (impl_1__new_const ([("with msg"%string : string)]))).

Record Foo_record : Type :=
  {
    Foo_f_field : t_u8;
  }.


#[export] Instance settable_Foo_record : Settable _ :=
  settable! (Build_Foo_record) <Foo_f_field>.

Instance t_StructuralPartialEq_61223753 : t_StructuralPartialEq ((t_Foo)) :=
  {
  }.

Instance t_PartialEq_662037409 : t_PartialEq ((t_Foo)) ((t_Foo)) :=
  {
    implaabbcc_t_PartialEq_f_eq := fun  (self : t_Foo) (other : t_Foo)=>
      f_eq (f_field self) (f_field other);
  }.

Instance t_Eq_795626685 : t_Eq ((t_Foo)) :=
  {
    implaabbcc_t_Eq_f_assert_receiver_is_total_eq := fun  (self : t_Foo)=>
      failure (("(AST import) something is not implemented yet.This is discussed in issue https://github.com/hacspec/hax/issues/156.
Please upvote or comment this issue if you see this error message.
Sorry, Hax does not support declare-first let bindings (see https://doc.rust-lang.org/rust-by-example/variable_bindings/declare.html) for now."%string : string)) ((""%string : string));
  }.

Definition v_CONSTANT : t_Foo :=
  Foo ((3 : t_u8)).

Definition numeric '(_ : unit) : unit :=
  let _ : t_usize := (123 : t_usize) in
  let _ : t_isize := (-42 : t_isize) in
  let _ : t_isize := (42 : t_isize) in
  let _ : t_i32 := (-42 : t_i32) in
  let _ : t_u128 := (22222222222222222222 : t_u128) in
  tt.

Definition patterns '(_ : unit) : unit :=
  let _ := match (1 : t_u8) with
  | 2 =>
    tt
  | _ =>
    tt
  end in
  let _ := match (("hello"%string : string),((123 : t_i32),[("a"%string : string); ("b"%string : string)])) with
  | ("hello"%string,(123,e_todo)) =>
    tt
  | _ =>
    tt
  end in
  let _ := match Foo ((4 : t_u8)) with
  | Foo (3) =>
    tt
  | _ =>
    tt
  end in
  tt.

Definition casts (x8 : t_u8) (x16 : t_u16) (x32 : t_u32) (x64 : t_u64) (xs : t_usize) : unit :=
  let _ : t_u64 := f_add (f_add (f_add (f_add (cast (x8)) (cast (x16))) (cast (x32))) (x64)) (cast (xs)) in
  let _ : t_u32 := f_add (f_add (f_add (f_add (cast (x8)) (cast (x16))) (x32)) (cast (x64))) (cast (xs)) in
  let _ : t_u16 := f_add (f_add (f_add (f_add (cast (x8)) (x16)) (cast (x32))) (cast (x64))) (cast (xs)) in
  let _ : t_u8 := f_add (f_add (f_add (f_add (x8) (cast (x16))) (cast (x32))) (cast (x64))) (cast (xs)) in
  let _ : t_i64 := f_add (f_add (f_add (f_add (cast (x8)) (cast (x16))) (cast (x32))) (cast (x64))) (cast (xs)) in
  let _ : t_i32 := f_add (f_add (f_add (f_add (cast (x8)) (cast (x16))) (cast (x32))) (cast (x64))) (cast (xs)) in
  let _ : t_i16 := f_add (f_add (f_add (f_add (cast (x8)) (cast (x16))) (cast (x32))) (cast (x64))) (cast (xs)) in
  let _ : t_i8 := f_add (f_add (f_add (f_add (cast (x8)) (cast (x16))) (cast (x32))) (cast (x64))) (cast (xs)) in
  tt.

Definition empty_array '(_ : unit) : unit :=
  let _ : t_Slice t_u8 := unsize ([]) in
  tt.

Definition fn_pointer_cast '(_ : unit) : unit :=
  let f : t_u32 -> t_u32 := fun x =>
    x in
  tt.

Definition null : ascii :=
  ("\000"%char : ascii).
'''
_CoqProject = '''
-R ./ TODO
-arg -w
-arg all

Literals.v'''
